crd/overlayextensionconfig/client.go

```diff
@@ -0,0 +1,119 @@
+type Installer struct {
+	cli typedv1.CustomResourceDefinitionInterface
+}
+
+func NewInstaller(c *rest.Config) (*Installer, error) {
+	cli, err := crd.NewCRDClientFromConfig(c)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to init crd client")
+	}
+	return &Installer{
+		cli: cli,
+	}, nil
+}
+
+func (i *Installer) create(ctx context.Context, res *v1.CustomResourceDefinition) (*v1.CustomResourceDefinition, error) {
+	res, err := i.cli.Create(ctx, res, metav1.CreateOptions{})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create oec crd")
+	}
+	return res, nil
+}
+
+// Install installs the embedded OverlayExtensionConfig CRD definition in the cluster.
+func (i *Installer) Install(ctx context.Context) (*v1.CustomResourceDefinition, error) {
+	oec, err := GetOverlayExtensionConfigs()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get embedded oec crd")
+	}
+	return i.create(ctx, oec)
+}
+
+// InstallOrUpdate installs the embedded OverlayExtensionConfig CRD definition in the cluster or updates it if present.
+func (i *Installer) InstallOrUpdate(ctx context.Context) (*v1.CustomResourceDefinition, error) {
+	oec, err := GetOverlayExtensionConfigs()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get embedded oec crd")
+	}
+	current, err := i.create(ctx, oec)
+	if !apierrors.IsAlreadyExists(err) {
+		return current, err
+	}
+	if current == nil {
+		current, err = i.cli.Get(ctx, oec.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get existing oec crd")
+		}
+	}
+	if !reflect.DeepEqual(oec.Spec.Versions, current.Spec.Versions) {
+		oec.SetResourceVersion(current.GetResourceVersion())
+		previous := *current
+		current, err = i.cli.Update(ctx, oec, metav1.UpdateOptions{})
+		if err != nil {
+			return &previous, errors.Wrap(err, "failed to update existing oec crd")
+		}
+	}
+	return current, nil
+}
+
+// Client provides methods to interact with instances of the OverlayExtensionConfig custom resource.
+type Client struct {
+	Cli client.Client
+}
+
+// NewClient creates a new OverlayExtensionConfig client around the passed ctrlcli.Client.
+func NewClient(cli client.Client) *Client {
+	return &Client{
+		Cli: cli,
+	}
+}
+
+func (c *Client) CreateOEC(ctx context.Context, oec *v1alpha1.OverlayExtensionConfig, fieldOwner string) error {
+	if err := c.Cli.Create(ctx, oec); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return errors.Wrap(err, "error creating OverlayExtensionConfig CRD")
+		}
+		if err := c.Cli.Patch(ctx, &v1alpha1.OverlayExtensionConfig{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: v1alpha1.GroupVersion.String(),
+				Kind:       "OverlayExtensionConfig",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      oec.Name,
+				Namespace: oec.Namespace,
+			},
+			Spec: oec.Spec,
+		}, client.Apply, client.ForceOwnership, client.FieldOwner(fieldOwner)); err != nil {
```

Comment:
who becomes the owner here?

Reply:
I assume AGIC controller will be an owner here as it would be the resource that created the OEC CR. I don't have a clear understanding of why we need this ownership though.

Reply:
suggest removing, just let DNC-RC's client.Client take care of doing what it wants

Reply:
Removed all of the Client code and left only the `InstallOrUpdate` method on the Installer